File: backend/src/services/card-effect.service.ts
import { GameState } from '../models/game-state';
import { CardEffect } from '../models/card-effect';
import { BattleEnemy } from '../models/battle-state';
import { Companion } from '../models/companion';

export type EffectSource = Companion | BattleEnemy;
export type EffectTarget = Companion | BattleEnemy;

/**
 * Single-responsibility service for applying CardEffects to game state.
 *
 * Rules:
 * - 'damage'     : reduce each target's HP through its shield first.
 *                  If a BattleEnemy reaches 0 HP and the source is a
 *                  Companion, set killedByCompanionId = source.id.
 * - 'shield'     : add shield value to the source (companions only).
 * - 'heal'       : restore HP to source companion up to maxLife.
 * - 'evade'      : no-op stub (future use).
 * - 'evade_draw' : no-op stub (future use).
 *
 * All methods return a new GameState — no in-place mutations.
 */
export class CardEffectService {
  apply(
    effect: CardEffect,
    source: EffectSource,
    targets: EffectTarget[],
    state: GameState
  ): GameState {
    switch (effect.action) {
      case 'damage':    return this.applyDamage(effect.value, source, targets, state);
      case 'shield':    return this.applyShield(effect.value, source, state);
      case 'heal':      return this.applyHeal(effect.value, source, state);
      case 'evade':
      case 'evade_draw':
      default:
        return state;
    }
  }

  // ---------------------------------------------------------------------------
  // Private helpers
  // ---------------------------------------------------------------------------

  private applyDamage(
    value: number,
    source: EffectSource,
    targets: EffectTarget[],
    state: GameState
  ): GameState {
    const targetIds          = new Set(targets.map(t => t.id));
    const isCompanionSource  = 'priceDecks' in source;

    const updatedEnemies = (state.battle?.enemies ?? []).map(enemy => {
      if (!targetIds.has(enemy.id)) return enemy;
      const dmgThrough = Math.max(0, value - enemy.shield);
      const newShield  = Math.max(0, enemy.shield - value);
      const newLife    = Math.max(0, enemy.life - dmgThrough);
      return {
        ...enemy,
        life:   newLife,
        shield: newShield,
        killedByCompanionId:
          newLife === 0 && !enemy.killedByCompanionId && isCompanionSource
            ? source.id
            : enemy.killedByCompanionId,
      };
    });

    const updatedCompanions = state.companions.map(companion => {
      if (!targetIds.has(companion.id)) return companion;
      const currentShield = (companion as any).shield ?? 0;
      const dmgThrough    = Math.max(0, value - currentShield);
      const newShield     = Math.max(0, currentShield - value);
      const newLife       = Math.max(0, companion.life - dmgThrough);
      return { ...companion, life: newLife, shield: newShield };
    });

    return {
      ...state,
      companions: updatedCompanions,
      battle: state.battle
        ? { ...state.battle, enemies: updatedEnemies }
        : state.battle,
    };
  }

  private applyShield(value: number, source: EffectSource, state: GameState): GameState {
    const updatedCompanions = state.companions.map(c =>
      c.id !== source.id
        ? c
        : { ...c, shield: ((c as any).shield ?? 0) + value }
    );
    return { ...state, companions: updatedCompanions };
  }

  private applyHeal(value: number, source: EffectSource, state: GameState): GameState {
    const updatedCompanions = state.companions.map(c => {
      if (c.id !== source.id) return c;
      const max = c.maxLife ?? c.life;
      return { ...c, life: Math.min(max, c.life + value) };
    });
    return { ...state, companions: updatedCompanions };
  }
}

File: backend-csharp/TheGameIsOn.API/Services/CardEffectService.cs
using TheGameIsOn.API.Models;

namespace TheGameIsOn.API.Services;

public interface ICardEffectService
{
    GameState Apply(CardEffect effect, IEffectEntity source, List<IEffectEntity> targets, GameState state);
}

/// <summary>
/// Single-responsibility service for applying CardEffects to game state.
/// All methods return a new GameState — no in-place mutations.
/// </summary>
public class CardEffectService : ICardEffectService
{
    public GameState Apply(CardEffect effect, IEffectEntity source, List<IEffectEntity> targets, GameState state)
    {
        return effect.Action switch
        {
            CardEffectAction.Damage => ApplyDamage(effect.Value, source, targets, state),
            CardEffectAction.Shield => ApplyShield(effect.Value, source, state),
            CardEffectAction.Heal   => ApplyHeal(effect.Value, source, state),
            // ApplyStatus is handled async by StatusEffectService in BattleService
            _ => state
        };
    }

    // -------------------------------------------------------------------------

    private GameState ApplyDamage(int value, IEffectEntity source, List<IEffectEntity> targets, GameState state)
    {
        var targetIds = new HashSet<string>(targets.Select(t => t.Id));
        var isCompanionSource = source is Companion;

        var updatedEnemies = (state.Battle?.Enemies ?? new()).Select(enemy =>
        {
            if (!targetIds.Contains(enemy.Id)) return enemy;

            var dmgThrough = Math.Max(0, value - enemy.Shield);
            var newShield  = Math.Max(0, enemy.Shield - value);
            var newLife    = Math.Max(0, enemy.Life - dmgThrough);

            return new BattleEnemy
            {
                Id                  = enemy.Id,
                DefinitionId        = enemy.DefinitionId,
                Name                = enemy.Name,
                Life                = newLife,
                MaxLife             = enemy.MaxLife,
                Shield              = newShield,
                Energy              = enemy.Energy,
                MaxEnergy           = enemy.MaxEnergy,
                Element             = enemy.Element,
                Type                = enemy.Type,
                Level               = enemy.Level,
                ExpReward           = enemy.ExpReward,
                Rewards             = enemy.Rewards,
                AttackSummaries     = enemy.AttackSummaries,
                KilledByCompanionId = newLife == 0 && enemy.KilledByCompanionId is null && isCompanionSource
                    ? source.Id
                    : enemy.KilledByCompanionId,
                StatusEffects       = enemy.StatusEffects
            };
        }).ToList();

        var updatedCompanions = state.Companions.Select(companion =>
        {
            if (!targetIds.Contains(companion.Id)) return companion;

            var currentShield = companion.Shield;
            var dmgThrough    = Math.Max(0, value - currentShield);
            var newShield     = Math.Max(0, currentShield - value);
            var newLife       = Math.Max(0, companion.Life - dmgThrough);

            return CloneCompanion(companion, c =>
            {
                c.Life   = newLife;
                c.Shield = newShield;
            });
        }).ToList();

        return new GameState
        {
            Player                = state.Player,
            Graph                 = state.Graph,
            Cards                 = state.Cards,
            Companions            = updatedCompanions,
            History               = state.History,
            Battle                = state.Battle is not null
                ? CloneBattle(state.Battle, b => b.Enemies = updatedEnemies)
                : null,
            PendingAbilityChoices = state.PendingAbilityChoices
        };
    }

    private GameState ApplyShield(int value, IEffectEntity source, GameState state)
    {
        var updatedCompanions = state.Companions.Select(c =>
            c.Id != source.Id
                ? c
                : CloneCompanion(c, comp => comp.Shield += value)
        ).ToList();

        return new GameState
        {
            Player                = state.Player,
            Graph                 = state.Graph,
            Cards                 = state.Cards,
            Companions            = updatedCompanions,
            History               = state.History,
            Battle                = state.Battle,
            PendingAbilityChoices = state.PendingAbilityChoices
        };
    }

    private GameState ApplyHeal(int value, IEffectEntity source, GameState state)
    {
        var updatedCompanions = state.Companions.Select(c =>
        {
            if (c.Id != source.Id) return c;
            var max = c.MaxLife ?? c.Life;
            return CloneCompanion(c, comp => comp.Life = Math.Min(max, c.Life + value));
        }).ToList();

        return new GameState
        {
            Player                = state.Player,
            Graph                 = state.Graph,
            Cards                 = state.Cards,
            Companions            = updatedCompanions,
            History               = state.History,
            Battle                = state.Battle,
            PendingAbilityChoices = state.PendingAbilityChoices
        };
    }

    // -------------------------------------------------------------------------
    // Shallow-clone helpers
    // -------------------------------------------------------------------------

    internal static Companion CloneCompanion(Companion src, Action<Companion>? mutate = null)
    {
        var c = new Companion
        {
            Id                 = src.Id,
            Name               = src.Name,
            Type               = src.Type,
            Element            = src.Element,
            Life               = src.Life,
            MaxLife            = src.MaxLife,
            Energy             = src.Energy,
            MaxEnergy          = src.MaxEnergy,
            EnergyRefill       = src.EnergyRefill,
            Sprite             = src.Sprite,
            PriceDecks         = src.PriceDecks,
            Level              = src.Level,
            Exp                = src.Exp,
            NextLevelExp       = src.NextLevelExp,
            AbilityUnlockLevels= new List<int>(src.AbilityUnlockLevels),
            AbilityPool        = new List<SpecialAbility>(src.AbilityPool),
            SpecialAbilities   = new List<SpecialAbility>(src.SpecialAbilities),
            Shield             = src.Shield,
            StatusEffects      = src.StatusEffects is not null ? new List<StatusEffect>(src.StatusEffects) : null
        };
        mutate?.Invoke(c);
        return c;
    }

    private static BattleState CloneBattle(BattleState src, Action<BattleState>? mutate = null)
    {
        var b = new BattleState
        {
            Active             = src.Active,
            Enemies            = new List<BattleEnemy>(src.Enemies),
            Turn               = src.Turn,
            Log                = new List<string>(src.Log),
            PendingCardRewards = new List<PendingCardReward>(src.PendingCardRewards),
            LastTurnActions    = src.LastTurnActions is not null ? new List<EnemyTurnAction>(src.LastTurnActions) : null
        };
        mutate?.Invoke(b);
        return b;
    }
}
